### PR TITLE
Fix involuntary cassandra upserts

### PR DESCRIPTION
**Pull request by fisx**
https://wearezeta.atlassian.net/browse/WPB-3915

## Checklist

 - [x] Re-visit all changed UPDATE commands, and see if UPSERT wasn't intentional.  If so, remove the `IF EXISTS`.
 - [x] Change types of commands that still contain `IF EXISTS` so they return `Row`, and fix type errors.
 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

**Comment by fisx** [comment]
lots of integration test failures, but i think they are unrelated?

**Comment by battermann** [comment]
I think there are a lot of places where we (primarily clients, but also in our integration tests) rely on the upsert semantic.

So I wouldn't say, that the upserts are necessarily all involuntary.

Even if we fix our integration test accordingly, we can't verify if those changes don't break clients.

I don't think we should proceed with this. Should we maybe rename all functions from `updateXY` to `upsertXY`.

**Comment by fisx** [comment]
> I don't think we should proceed with this. Should we maybe rename all functions from `updateXY` to `upsertXY`.

good idea!

> So I wouldn't say, that the upserts are necessarily all involuntary.

i've worked under the assumption that if the update would create an opviously broken row in case of insert, then it's probably a proper update, not an upsert.  Is that rule wrong, or did I not apply it carefully enough, or are the test failures unrelated?

(note that i *did* spot a few upserts, and fixed them in the last few commits.)

**Comment by battermann** [comment]
> i've worked under the assumption that if the update would create an opviously broken row in case of insert, then it's probably a proper update, not an upsert. Is that rule wrong, or did I not apply it carefully enough, or are the test failures unrelated?

Yes, that makes sense. If an insert would lead to corrupted state, that should be the right place to check for existence. The test failures I saw were indeed unrelated, that's why I was more concerned than necessary.

File: services/brig/src/Brig/Data/Connection.hs [hunk 1]
```diff
@@ -339,8 +341,8 @@ deleteRemoteConnectionsDomain dom = do
 connectionInsert :: PrepQuery W (UserId, UserId, RelationWithHistory, UTCTimeMillis, ConvId) ()
 connectionInsert = "INSERT INTO connection (left, right, status, last_update, conv) VALUES (?, ?, ?, ?, ?)"
 
-connectionUpdate :: PrepQuery W (RelationWithHistory, UTCTimeMillis, UserId, UserId) ()
-connectionUpdate = "UPDATE connection SET status = ?, last_update = ? WHERE left = ? AND right = ?"
+connectionUpdate :: PrepQuery W (RelationWithHistory, UTCTimeMillis, UserId, UserId) Row
```
Comment (battermann) [inline services/brig/src/Brig/Data/Connection.hs:344]: Yes, update without convId does not make sense.

File: services/galley/src/Galley/Cassandra/Client.hs [hunk 2]
```diff
@@ -42,7 +42,7 @@ import UnliftIO qualified
 updateClient :: Bool -> UserId -> ClientId -> Client ()
 updateClient add usr cls = do
   let q = if add then Cql.addMemberClient else Cql.rmMemberClient
-  retry x5 $ write (q cls) (params LocalQuorum (Identity usr))
+  retry x5 . void $ trans (q cls) (params LocalQuorum (Identity usr))
```
Comment (battermann) [inline services/galley/src/Galley/Cassandra/Client.hs:45]: Not sure here for `addMemberClient`. Is it possible that a user has no clients, and this really is an upsert?

File: services/galley/src/Galley/Cassandra/Queries.hs [hunk 3]
```diff
@@ -443,15 +443,15 @@ selectClients = "select user, clients from clients where user in ?"
 rmClients :: PrepQuery W (Identity UserId) ()
 rmClients = "delete from clients where user = ?"
 
-addMemberClient :: ClientId -> QueryString W (Identity UserId) ()
+addMemberClient :: ClientId -> QueryString W (Identity UserId) Row
```
Comment (battermann) [inline services/galley/src/Galley/Cassandra/Queries.hs:450]: Is it possible that a user has no clients, and no entry exists?

File: services/gundeck/src/Gundeck/Push/Data.hs [hunk 4]
```diff
@@ -49,10 +49,10 @@ insert u t a p e o c = retry x5 $ write q (params LocalQuorum (u, t, a, p, e, o,
     q = "insert into user_push (usr, transport, app, ptoken, arn, connection, client) values (?, ?, ?, ?, ?, ?, ?)"
 
 updateArn :: MonadClient m => UserId -> Transport -> AppName -> Token -> EndpointArn -> m ()
-updateArn uid transport app token arn = retry x5 $ write q (params LocalQuorum (arn, uid, transport, app, token))
+updateArn uid transport app token arn = retry x5 . void $ trans q (params LocalQuorum (arn, uid, transport, app, token))
```
Comment (battermann) [inline services/gundeck/src/Gundeck/Push/Data.hs:52]: I guess ARN without connectionId and clientId is not valid?

File: tools/db/migrate-sso-feature-flag/src/Work.hs [hunk 5]
```diff
@@ -66,6 +66,7 @@ writeSsoFlags = mapM_ (`setSSOTeamConfig` FeatureStatusEnabled)
   where
     setSSOTeamConfig :: MonadClient m => TeamId -> FeatureStatus -> m ()
     setSSOTeamConfig tid ssoTeamConfigStatus = do
-      retry x5 $ write updateSSOTeamConfig (params LocalQuorum (ssoTeamConfigStatus, tid))
-    updateSSOTeamConfig :: PrepQuery W (FeatureStatus, TeamId) ()
-    updateSSOTeamConfig = "update team_features set sso_status = ? where team_id = ?"
+      retry x5 . void $ trans updateSSOTeamConfig $ params LocalQuorum (ssoTeamConfigStatus, tid)
```
Comment (battermann) [inline tools/db/migrate-sso-feature-flag/src/Work.hs:69]: Not sure here either?

**Review by battermann — COMMENTED** [review]
I made some comments where I am not sure, please review these. Otherwise it looks good!

File: services/galley/src/Galley/Cassandra/Client.hs (hunk 2 above)
Comment (fisx) [reply services/galley/src/Galley/Cassandra/Client.hs:45]: not sure, you may be right 52d4470f6

Comment (fisx) [reply services/galley/src/Galley/Cassandra/Client.hs]: sorry, i missed a spot: ea8595e3a

File: services/gundeck/src/Gundeck/Push/Data.hs (hunk 4 above)
Comment (fisx) [reply services/gundeck/src/Gundeck/Push/Data.hs:52]: i would think so, yes.

File: tools/db/migrate-sso-feature-flag/src/Work.hs (hunk 5 above)
Comment (fisx) [reply tools/db/migrate-sso-feature-flag/src/Work.hs:69]: I'm pretty sure we don't want to configure deleted or never-created teams, so I think this is fine.  But if you don't believe me I'll change it back, this is almost-dead code anyway.  :)

File: services/galley/src/Galley/Cassandra/CustomBackend.hs [hunk 6]
```diff
@@ -51,7 +51,7 @@ getCustomBackend domain =
 
 setCustomBackend :: MonadClient m => Domain -> CustomBackend -> m ()
 setCustomBackend domain CustomBackend {..} = do
-  retry x5 $ write Cql.updateCustomBackend (params LocalQuorum (backendConfigJsonUrl, backendWebappWelcomeUrl, domain))
+  retry x5 . void $ trans Cql.updateCustomBackend (params LocalQuorum (backendConfigJsonUrl, backendWebappWelcomeUrl, domain))
```
[the remark this answers is not in the document]
Comment (fisx) [reply services/galley/src/Galley/Cassandra/CustomBackend.hs]: you're right.  damn!  0494cd332

**Comment by fisx** [comment]
got caught by @akshaymankar and @pcapriotti: 'if exists' is dangerous for cassandra performance!

so this PR is less exciting than i was hoping it would get, but at least we have more docs in the code now, so this effort won't likely be spent again in the future without effect.

**Review by battermann — APPROVED** [review]

**Pull request merged**